cli/tests/default/unit/test_engine_type.py

```diff
@@ -2,79 +2,127 @@
+        # semgrep ci with toggle on, diff scan
+        (True, True, False, False, ET.PRO_INTRAFILE),
+        (True, True, False, True, ET.PRO_INTERFILE),
         # semgrep ci with toggle off, diff scan
-        (False, False, False, None, ET.OSS),
-        (False, False, False, ET.OSS, ET.OSS),
-        (False, False, False, ET.PRO_LANG, ET.PRO_LANG),
-        (False, False, False, ET.PRO_INTRAFILE, ET.PRO_INTRAFILE),
-        (False, False, False, ET.PRO_INTERFILE, ET.PRO_INTRAFILE),
-        (False, False, True, ET.PRO_INTERFILE, ET.PRO_INTERFILE),
+        (True, False, False, False, ET.PRO_INTRAFILE),
+        (True, False, False, True, ET.PRO_INTRAFILE),
     ],
 )
 def test_decide_engine_type(
-    mocker, is_cloud_flag_on, is_ci_scan_full, enable_pro_diff, requested, expected
+    mocker,
+    logged_in,
+    is_interfile_flag_on,
+    is_ci_scan_full,
+    interfile_diff_scan_enabled,
+    is_supply_chain_only,
+    is_secrets_scan,
+    engine_flag,
+    expected_default,
 ):
     scan_handler = None
     git_meta = None
 
-    if is_ci_scan_full is not None:  # None means we're in `semgrep scan`
+    if is_interfile_flag_on is not None:  # None means we're in `semgrep scan`
         scan_handler = mocker.Mock(spec=ScanHandler)
-        scan_handler.deepsemgrep = is_cloud_flag_on
+        scan_handler.deepsemgrep = is_interfile_flag_on
 
+    if is_ci_scan_full is not None:  # None means there was no metadata
         git_meta = mocker.Mock(spec=GitMeta)
         git_meta.is_full_scan = is_ci_scan_full
 
-    assert (
-        ET.decide_engine_type(
-            requested_engine=requested,
-            scan_handler=scan_handler,
-            git_meta=git_meta,
-            enable_pro_diff_scan=enable_pro_diff,
+    args = [
+        logged_in,
+        engine_flag,
+        is_secrets_scan,
+        interfile_diff_scan_enabled,
+        scan_handler,
+        git_meta,
+        is_supply_chain_only,
+    ]
+    if is_secrets_scan and engine_flag is ET.OSS:
+        pytest.raises(SemgrepError, ET.decide_engine_type, *args)
+    else:
+        diff_scan_override = not (
+            (is_ci_scan_full is None or is_ci_scan_full) or interfile_diff_scan_enabled
         )
-        == expected
-    )
-
-    # Expect engine to be non-interfile/intrafile pro
-    expected_supply_chain_only_engine = expected if expected == ET.OSS else ET.PRO_LANG
-    assert (
-        ET.decide_engine_type(
-            requested_engine=requested,
-            scan_handler=scan_handler,
-            git_meta=git_meta,
-            enable_pro_diff_scan=enable_pro_diff,
-            supply_chain_only=True,
+        assert ET.decide_engine_type(*args) == expected_engine_type(
+            is_supply_chain_only,
+            is_secrets_scan,
+            diff_scan_override,
+            engine_flag,
+            expected_default,
         )
-        == expected_supply_chain_only_engine
-    )
+
+
+def expected_engine_type(
```

Comment:
That we need a function to compute the expected engine type doesn't feel right. Or am I understanding it the wrong way?

Reply:
You're understanding it correctly. I thought about a few options, but I found that this was the simplest way to write the unit tests. With seven arguments, enumerating every possible combination is impractical, but I wanted thorough test coverage here because of how hard `engine.py` is to understand.